routes/login.js

```diff
@@ -0,0 +1,142 @@
+const router = require("express").Router();
+const bcrypt = require("bcryptjs");
+const jwtService = require("../services/JwtService");
+const Error = require("../models/Error");
+const Joi = require("@hapi/joi");
+
+const userService = require("../services/UserService");
+
+router.post("/login", (req, res) => {
+    const { zprn, password } = req.body;
+    const { error, value } = validateCredentials(zprn, password);
+    if (error != undefined) {
+        res.status(400).send(
+            new Error("BAD_REQUEST", error.details[0].message)
+        );
+    } else {
+        userService
+            .getUser(zprn)
+            .then((user) => {
+                bcrypt.compare(password, user.password, (error, match) => {
+                    if (error) {
+                        res.status(500).send(
+                            new Error("INTERNAL_SERVER_ERROR", error)
+                        );
+                    } else {
+                        if (match) {
+                            const token = jwtService.sign(user.department);
+                            token != null
+                                ? res.status(200).json({ token: token })
+                                : res
+                                      .status(500)
+                                      .send(
+                                          new Error(
+                                              "INTERNAL_SERVER_ERROR",
+                                              "Internal error occurred!"
+                                          )
+                                      );
+                        } else {
+                            res.status(401).send(
+                                new Error("UNAUTHORIZED", "Password incorrect!")
+                            );
+                        }
+                    }
+                });
+            })
+            .catch(() => {
+                res.status(400).send(
+                    new Error("BAD_REQUEST", "User not found!")
+                );
+            });
+    }
+});
+
+router.post("/add", (req, res) => {
+    const { zprn, firstName, lastName, password, department, role } = req.body;
+    const { er, value } = validateData(
+        zprn,
+        firstName,
+        lastName,
+        password,
+        department,
+        role
+    );
+    if (er != undefined) {
+        res.status(400).send(
+            new Error("BAD_REQUEST", error.details[0].message)
+        );
+    } else {
+        bcrypt.genSalt(10, (e, salt) => {
+            if (!e) {
+                bcrypt.hash(password, salt, (error, hash) => {
+                    if (error) {
+                        res.status(500).send(
+                            new Error("INTERNAL_SERVER_ERROR", error)
+                        );
+                    } else {
+                        if (!hash) {
```

Comment:
Do we have to check for `!hash` as, if there is an error we are already catching it above in `if (error)` right? So we already know that the hash is not going to be empty.

Reply:
Agree on this, I just did it to be on a safer side. Would surely remove it!